fix(settings): show active forced locale or language instead of 'No locale set' (#41543)

arc64 opened the pull request:
* Fixes #41543

## Summary

When a forced locale or language is set by the administrator, this PR ensures the appropriate message is displayed instead of the misleading 'No locale set'.

1. Updated getLanguageMap() and getLocaleMap() in PersonalInfo.php to pass through validated forced settings. If settings are malformed, no forced value is passed (falls back to default). 
2. Modified LanguageSection.vue and LocaleSection.vue to correctly handle display conditions:

- If forcedLocale or forcedLanguage exists, show 'Locale is forced to {locale} by the administrator' or 'Language is forced to {language} by the administrator'.
- If no forced value and no user settings are present, it falls back to 'No locale set' / 'No language set'.

3. Added tests for both Vue components to ensure correct behaviour

## Before

[Issue #41543 - Screenshot](https://github.com/nextcloud/server/issues/41543)

## After

**Forced locale / language:**

<img width="718" alt="Forced" src="https://github.com/user-attachments/assets/ae687523-ef20-4e6d-98db-3430045e0e28" />

**User-set locale / language:**

<img width="737" alt="Set by user" src="https://github.com/user-attachments/assets/ed0a54a5-cfaf-4307-b117-00ed6f97065b" />


## TODO

- [ ]  Check if PHP unit tests are required for getLanguageMap() and getLocaleMap(). I have not written tests yet, as running them in the current docker dev container was not successful (due to missing full test environment / FS setup). Feedback welcome.
- [ ] Check that the messaging in the views is inline with Nextcloud expectations - 'Locale is forced to {locale} by the administrator' and 'Language is forced to {language} by the administrator'. 

## Notes
When the locale is set (either by the user or forced by an administrator), it is displayed in English, regardless of the current interface language. This matches existing behaviour in user settings.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

arc64 committed:
fix(settings): show active forced locale or language instead of 'No locale set' (fixes #41543)

Signed-off-by: Annabel Church <215145+arc64@users.noreply.github.com>

susnux commented on apps/settings/lib/Settings/Personal/PersonalInfo.php:
```diff
@@ -261,9 +284,32 @@ private function getLanguageMap(IUser $user): array {
+	 * Validates a forced locale setting against available locales
+	 */
+	private function validateForcedLocale(string $forcedLocale, array $localeCodes): ?array {
+		$forcedLocaleObj = array_filter($localeCodes, fn($locale) => $locale['code'] === $forcedLocale);
```
Thats not an object the double usage of the name if often considered confusing.

susnux commented on apps/settings/src/components/PersonalInfo/LanguageSection/__tests__/LanguageSection.spec.js:
We usually place tests next to the code in server, like this would be
`apps/settings/src/components/PersonalInfo/LanguageSection/LanguageSection.spec.js`

arc64 replied:
Thank you, I'll get them into the right place

susnux commented on apps/settings/src/components/PersonalInfo/LanguageSection/__tests__/LanguageSection.spec.js:
```diff
@@ -0,0 +1,152 @@
+/**
+ * Mock Nextcloud modules
+ */
+vi.mock('../../../constants/AccountPropertyConstants.js', () => ({
+	ACCOUNT_SETTING_PROPERTY_ENUM: { LANGUAGE: 'language' },
+	ACCOUNT_SETTING_PROPERTY_READABLE_ENUM: { LANGUAGE: 'Language' },
+}))
```
Why do you need to mock constants?

arc64 replied:
I removed this

susnux commented on apps/settings/src/components/PersonalInfo/LanguageSection/__tests__/LanguageSection.spec.js:
```diff
@@ -0,0 +1,152 @@
+	})),
+}))
+
+vi.mock('@nextcloud/l10n', () => ({
+	t: (app, text, params) => {
+		if (params) {
+			return text.replace(/\{(\w+)\}/g, (match, key) => params[key] || match)
+		}
+		return text
+	},
+	getLanguage: () => 'en',
+	isRTL: () => false,
+	translate: (app, text, params) => {
+		if (params) {
+			return text.replace(/\{(\w+)\}/g, (match, key) => params[key] || match)
+		}
+		return text
+	},
+}))
```
I do not think you need to mock this.

arc64 replied:
I've removed these

susnux commented on apps/settings/src/components/PersonalInfo/LanguageSection/__tests__/LanguageSection.spec.js:
```diff
@@ -0,0 +1,152 @@
+			const emptyData = { languageMap: {} }
+			vi.mocked(loadState).mockReturnValueOnce(emptyData)
+			wrapper = mountComponent()
+			await wrapper.vm.$nextTick()
+			expect(wrapper.vm.isEditable).toBe(false)
+			expect(wrapper.vm.language).toBeNull()
+			expect(wrapper.vm.commonLanguages).toEqual([])
+			expect(wrapper.vm.otherLanguages).toEqual([])
+			expect(wrapper.vm.forcedLanguage).toBeNull()
```
I do not think testing internals is a good thing its like testing private methods in classes.

arc64 replied:
I've removed the reliance on checking the text for the components and have moved to using data-test attributes.

susnux commented on apps/settings/src/components/PersonalInfo/LanguageSection/__tests__/LanguageSection.spec.js:
Nice that there is a test but for this its probably also good to add a cypress e2e test.

pringelmann replied:
@arc64 did you manage to get the cypress tests working? What specific issue are you experiencing?

susnux commented on apps/settings/src/components/PersonalInfo/LocaleSection/__tests__/LocaleSection.spec.js:
Similar like the language one

arc64 replied:
I've modified this too

susnux reviewed (commented):
Nice, thank you for the pull request! Added some comments :)

arc64 committed:
fix(settings): improve forced locale/language validation and testability

- Clarified the naming in the validate functions
- Removed unnecessary mocks, and added stubs in tests.
- Testing for existence of element with data-test attributes, and without relying on translated text.
- Removed overly complex component mocking

Signed-off-by: Annabel Church <215145+arc64@users.noreply.github.com>

arc64 commented:
I forgot to re-request a review, done now, @susnux thank you for your comments and labour
